lib/issued_cid.c

```diff
@@ -0,0 +1,168 @@
+ */
+static void do_mark_pending(quicly_issued_cid_set_t *set, size_t idx)
+{
+    set->cids[idx].state = QUICLY_ISSUED_CID_STATE_PENDING;
+    for (size_t j = 0; j < idx; j++) {
+        if (set->cids[j].state != QUICLY_ISSUED_CID_STATE_PENDING) {
+            swap_cids(&set->cids[idx], &set->cids[j]);
+            break;
+        }
+    }
+}
+
+void quicly_issued_cid_set_capacity(quicly_issued_cid_set_t *set, size_t capacity)
+{
+    assert(capacity >= 0);
+    assert(capacity <= PTLS_ELEMENTSOF(set->cids));
+    assert(set->_capacity <= capacity);
+
+    for (size_t i = set->_capacity; i < capacity; i++)
+        set->cids[i].state = QUICLY_ISSUED_CID_STATE_IDLE;
+
+    set->_capacity = capacity;
+
+    /* First we prepare N CIDs (to be precise here we prepare N-1, as we already had one upon initialization).
+     * Later, every time one of the CIDs is retired, we immediately prepare one additional CID
+     * to always fill the CID list. */
+    for (size_t i = 0; i < capacity; i++) {
+        if (set->cids[i].state != QUICLY_ISSUED_CID_STATE_IDLE)
+            continue;
+
+        if (set->_generator == NULL || set->_generator(set->_conn, &set->cids[i]) != 0)
+            break;
+        do_mark_pending(set, i);
+    }
+}
+
+void quicly_issued_cid_mark_inflight(quicly_issued_cid_set_t *set, size_t num_sent)
+{
+    assert(num_sent <= set->_capacity);
+
+    /* first, mark the first `num_sent` CIDs as INFLIGHT */
+    for (size_t i = 0; i < num_sent; i++) {
+        assert(set->cids[i].state == QUICLY_ISSUED_CID_STATE_PENDING);
+        set->cids[i].state = QUICLY_ISSUED_CID_STATE_INFLIGHT;
+    }
+
+    /* then move the remaining PENDING CIDs (if any) to the front of the array */
+    for (size_t i = num_sent; i < set->_capacity; i++) {
+        if (set->cids[i].state != QUICLY_ISSUED_CID_STATE_PENDING)
+            break;
+        swap_cids(&set->cids[i], &set->cids[i - num_sent]);
+    }
+}
+
+static size_t find_inflight(const quicly_issued_cid_set_t *set, uint64_t sequence)
+{
+    for (size_t i = 0; i < set->_capacity; i++) {
+        if (set->cids[i].sequence == sequence) {
+            assert(set->cids[i].state != QUICLY_ISSUED_CID_STATE_PENDING);
+            assert(set->cids[i].state != QUICLY_ISSUED_CID_STATE_IDLE);
+            return i;
+        }
+    }
+
+    return SIZE_MAX;
+}
+
+int quicly_issued_cid_mark_delivered(quicly_issued_cid_set_t *set, uint64_t sequence)
+{
+    size_t i = find_inflight(set, sequence);
+    if (i == SIZE_MAX)
+        return 1;
+
+    assert(set->cids[i].state == QUICLY_ISSUED_CID_STATE_INFLIGHT);
+
+    set->cids[i].state = QUICLY_ISSUED_CID_STATE_DELIVERED;
+
+    return 0;
+}
+
+int quicly_issued_cid_mark_pending(quicly_issued_cid_set_t *set, uint64_t sequence)
+{
+    size_t i = find_inflight(set, sequence);
+    if (i == SIZE_MAX)
+        return 1;
+
+    do_mark_pending(set, i);
+
+    return 0;
+}
```

Comment:
I'm not sure if the logic in `quicly_issued_cid_mark_delivered` and `quicly_issued_cid_mark_pending` are correct.

When the endpoint receives a late ack, an entry marked as PENDING can transition to DELIVERED. Similarly, a packet (and therefore a frame) can be declared lost after another packet carrying the same information was being acked.

It might be a good idea to rename these functions to "_acked", "_lost" at the same time changing the logic to handle these cases. The reason I suggest renaming is because the latter "mark_pending," is because, as stated, a frame deemed lost does not necessarily mean that it would be remarked as pending.